(6/6) Clean up technical debt in P2PDataStorage and ProtectedStorageEntry objects

julianknutsen opened the pull request:
b6b0026

## Motivation
This PR cleans up technical debt in the P2PDataStorage and the ProtectedStorageEntry objects that should be done before new features are added.

Future patches will change these objects quite a bit and getting them into a state that is easy to reason about will make the reviews easier.

## Testing
All existing unit tests continue to pass

julianknutsen committed:
[TESTS] Add tests of requestData

This will allow us to push the GetData creation inside P2PDataStorage
safely.

julianknutsen committed:
[REFACTOR] Introduce buildGetDataRequest variants

As part of changing the GetData path, we want to move all creation
and processing of GetData messages inside P2PDataStorage. This will allow
easier unit testing of the behavior as well as cleaner code in the
request handlers that can just focus on nonces, connections, etc.

julianknutsen committed:
[TESTS] Add tests of new RequestData APIs

These are identical test cases to the requestHandler tests, but with much
fewer dependencies. The requestHandler tests will eventually be deleted,
but they are going to remain throughout development as an extra safety
net.

julianknutsen committed:
[TESTS] Add tests of GetDataRequestHandler

Add some basic sanity tests prior to the refactor to help catch issues.

julianknutsen committed:
[REFACTOR] Introduce buildGetDataResponse

This is just a strict move of code to reduce errors.

julianknutsen committed:
[REFACTOR] Extract connectionInfo String

julianknutsen committed:
[REFACTOR] Extract getDataResponse logging

Changed the log to reference getDataResponse instead of getData. Now
that we might truncate the response, it ins't true that this is exactly
what the peer asked.

julianknutsen committed:
[REFACTOR] Extract truncation logging

Move the logging that utilizes connection information into the request
handler. Now, buildGetDataResponse just returns whether or not the list
is truncated which will make it easier to test.

julianknutsen committed:
[REFACTOR] Pass peerCapabilities into buildGetDataResponse

Remove the dependence on the connection object by having the handler
pass in the peer's capabilities. This now allows unit testing of
buildGetDataResponse without any connection dependencies.

julianknutsen committed:
[TESTS] Unit tests of buildGetDataResponse

Write a full set of unit tests for buildGetDataResponse. This provides
a safety net with additional refactoring work.

julianknutsen committed:
Remove redundant HashSet lookups in filter functions

The appendOnlyDataStoreService and map already have unique keys that
are based on the hash of the payload. This would catch instances
where:

PersistableNetworkPayload
- None: The key is based on ByteArray(payload.getHash()) which is the
        same as this check.

ProtectedStorageEntry
- Cases where multiple PSEs contain payloads that have equivalent
  hashCode(), but different data.toProtoMessage().toByteArray().
  I don't think it is a good idea to keep 2 "unique" methods on
  payloads. This is likely left over from a time when
  Payload hashCode() needed to be different than the hash of
  the payload.

julianknutsen committed:
[REFACTOR] Move required capabilities log

Move the logging function to the common capabilities check
so it can run on both ProtectedStoragePayload and
PersistableNetworkPayload objects

julianknutsen committed:
[REFACTOR] Inline capability check for ProtectedStorageEntries

Move the capability check inside the stream operation. This should
improve performance slightly, but more importantly it makes the
two filter functions almost identical so they can be combined.

julianknutsen committed:
[REFACTOR] Inline filtering functions

Removes unnecessary calculations converting Set<byte[]> into
Set<ByteArray> and allows additional deduplication of stream operations.

julianknutsen committed:
[REFACTOR] Remove duplication in filtering functions

Introduce a generic function that can be used to filter
Map<ByteArray, PersistableNetworkPayload> or
Map<ByteArray, ProtectedStorageEntry>.

Used to deduplicate the GetData code paths and ensure the logic is the
same between the two payload types.

julianknutsen committed:
[BUGFIX] Fix off-by-one in truncation logic

Now, the truncation is only triggered if more than MAX_ENTRIES could
have been returned.

julianknutsen committed:
[TESTS] Add test of RequestDataHandler::onMessage

Add heavy-handed test that exercises the logic to use as a safeguard
for refactoring.

julianknutsen committed:
[REFACTOR] Introduce processGetDataResponse

Just a code move for now.

julianknutsen committed:
[TESTS] Make verify() functions more flexible

Now that we want to unit test the GetData path which has different
behavior w.r.t. broadcasts, the tests need a way to verify that
state was updated, but not broadcast during an add.

This patch changes all verification function to take each state update
explicitly so the tests can do the proper verification.

julianknutsen committed:
[TESTS] Add unit tests for processGetDataResponse

Add a full set of unit tests that uncovered some unexpected
behavior w.r.t. signalers.

julianknutsen committed:
Remove static from initialRequestApplied

Previously, multiple handlers needed to signal off one global variable.
Now, that this check is inside the singleton P2PDataStorage, make it
non-static and private.

julianknutsen committed:
[TESTS] Write synchronization integration tests

Write a few integration test that exercises the exercise interesting
synchronization states including the lost remove bug. This fails
with the proper validation, but will pass at the end of the new feature
development.

julianknutsen committed:
[REFACTOR] Clean up ClientAPI for addPersistableNetworkPayload

Now that more callers have moved internal, the public facing API
can be cleaner and more simple. This should lead to a more maintainable
API and less sharp edges with future development work.

julianknutsen committed:
[REFACTOR] Clean up ClientAPI for addProtectedStorageEntry

Remove isDataOwner from the client API. All users pass in true. All test
users don't care.

julianknutsen committed:
[REFACTOR] Clean up ClientAPI for remove

Remove isDataOwner from the client API. All users pass in true. All test
users don't care.

julianknutsen committed:
[REFACTOR] Clean up ClientAPI for refreshTTL

Remove isDataOwner from the client API. All users pass in true. All test
users don't care.

julianknutsen committed:
[REFACTOR] inline broadcast() private function

Now that it is identical the the broadcaster version it can be
inlined.

julianknutsen committed:
[REFACTOR] inline broadcastProtectedStorageEntry() private function

Now that it is identical the the broadcaster version it can be
inlined.

julianknutsen committed:
[REFACTOR] inline maybeAddToRemoveAddOncePayloads() private function

Now that there is only one user it can be inlined.

julianknutsen committed:
[REFACTOR] Clean up processGetDataResponse

- Add more comments
- Use Clock instead of System
- Remove unnecessary AtomicInteger

julianknutsen committed:
[RENAME] LazyProcessedPayload to ProcessOncePersistableNetworkPayload

Name is left over from previous implementation. Change it to be more
relevant to the current code and update comments to indicate the
current usage.

julianknutsen committed:
Remove @Nullable around persistableNetworkPayloadSet

Checking for null creates hard-to-read code and it is simpler to just
create an empty set if we receive a pre-v0.6 GetDataResponse protobuf
message that does not have the field set.

julianknutsen committed:
Remove @Nullable around supportedCapabilities in PreliminaryGetDataRequest

The only two users of this constructor are the fromProto path which
now creates an empty Capabilities object similar to GetDataResponse.
The other internal usage of Capabilities.app which is initialized to empty.

julianknutsen committed:
[DEADCODE] Remove old request handler tests

Now that all the implementations are unit tested in P2PDataStorage,
the old tests can be removed.

julianknutsen committed:
Remove @Nullable around supportedCapabilities in GetDataResponse

The only two users of this constructor are the fromProto path which
already creates an empty Capabilities object if one is not provided and
the internal usage of Capabilities.app which is initialized to empty.

Remove the @Nullable so future readers aren't confused.

julianknutsen committed:
Make isDataOwner a private policy decision in BroadcastHandler

isDataOwner is used when deciding how many peer nodes should receive
a BroadcastMessage. If the BroadcastMessage originated
on the local node it is sent to ALL peer nodes with a small delay.

If the node is only relaying the message (it originated on a different
node) it is sent to MAX(peers.size(), 7) peers with a delay that is
twice as long.

All the information needed to determine whether or not the
BroadcastMessage originated on the local node is available at the final
broadcast site and there is no reason to have callers pass it in.

In the event that the sender address is not known during broadcast (which
is only a remote possibility due to how early the local node address
is set during startup) we can default to relay mode.

This first patch just removes the deep parameters. The next will remove
everything else. There is one real change in LiteNodeNetworkService.java
where it was using the local node when it should have been using the
peer node. This was updated to the correct behavior.

julianknutsen committed:
Remove isDataOwner from P2PDataStorage

Remove all usages in code and tests now that the behavior is
internal to the BroadcastHandler

julianknutsen committed:
[TESTS] Clean up mockito never() and eq(null) usages

never() and any() don't play well together for nullable types. Change
the broadcast mocks to user nullable() and fixup tests.

julianknutsen committed:
Make addPersistableNetworkPayloadFromInitialRequest private

Now that the only user is internal, the API can be made private and the
tests can be removed. This involved adding a few test cases to
processGetDataResponse to ensure the invalid hash size condition was
still covered.

julianknutsen committed:
Remove ProtectedStorageEntry::updateSignature

The only users were tests that can just pass a bad signature directly
into the constructor.

julianknutsen committed:
Remove ProtectedStorageEntry::maybeAdjustCreationTimeStamp

There is only 1 caller can be replaced with Math.min()

julianknutsen committed:
@NotNull ProtectedStorageEntry::ownerPubKey

In proto3 the initialized value is an empty ByteString and there are no
valid uses of passing in null here.

julianknutsen committed:
@NotNull ProtectedStorageEntry::protectedStoragePayload

The ProtectedStoragePayload.fromProto code will throw an exception if
this is null from the wire so there is no valid use for it to be null.

julianknutsen committed:
@NotNull MailboxStoragePayload::senderPubKeyForAddOperation

In proto3 this is intialized to an empty ByteString so there is no valid
use for it to be null.

julianknutsen committed:
Make CHECK_TTL_INTERVAL_SEC final

It is never changed

julianknutsen committed:
s/networkPayload/protectedStoragePayload

Helps readability when the variable name matches the type.

julianknutsen committed:
[TESTS] Make onDisconnect tests more robust

Before refactoring the function ensure the tests cover all cases. This
fixes a bug where the payload ttl was too low in some instances causing
backDate to do no work when it should.

julianknutsen committed:
Refactor P2PDataStorage::onDisconnect

1. Remove delete during stream iteration
2. Minimize branching w/ early returns for bad states
3. Use stream filter for readability
4. Implement additional checks that should be done when removing entries

julianknutsen committed:
Remove expire optimization in onDisconnect

We already have a garbage collection thread that runs every minute
to clean up items. Doing it again during onDisconnect is an unnecessary
optimization that adds complexity and caused bugs.

For example, the original implementation did not handle the sequence
number map correctly and was removing entries during a stream iteration.

This also reduces the complexity of testing. There is one code path
responsible for reducing ttls and one code path responsible for
expiring entries. Much easier to reason about.

julianknutsen committed:
Remove filter for ExpirablePayload

ProtectedStorageEntry::backDate() already handles this

freimair commented on p2p/src/main/java/bisq/network/p2p/storage/payload/ProtectedMailboxStorageEntry.java:
```diff
@@ -107,7 +107,7 @@ public boolean isValidForAddOperation() {
             String res1 = this.toString();
             String res2 = "null";
-            if (mailboxStoragePayload != null && mailboxStoragePayload.getOwnerPubKey() != null)
+            if (mailboxStoragePayload.getOwnerPubKey() != null)
```
(applies to all removed null-checks)
you say in the commit comment that `fromProto` will throw an exception. I cannot see that. However, it seems like the protobuf implementation does supply a default value which is not null. So either way this change should be fine. Or did I miss something here?

julianknutsen replied:
I could have been more clear with the distinction. Here is the relevant code for the payload itself. Since it is a submessage in proto3 it can technically be null, but the resolver handles those cases and raises an exception:

https://github.com/bisq-network/bisq/blob/76e8c5736a6c5f56cca83323c373d2f3eb3f2542/core/src/main/java/bisq/core/proto/network/CoreNetworkProtoResolver.java#L249-L278

The other non-null issues were due to the initialized value for proto3 not being null, this isn't one of those cases.

freimair replied:
ah damn. missed that.

freimair reviewed (dismissed):
Ack

ripcurlx commented:
> Ack

@freimair Could you please add how you tested this code changes, so it easier to compare with other ACKs. Thanks!

ripcurlx commented on p2p/src/main/java/bisq/network/p2p/storage/P2PDataStorage.java:
```diff
@@ -440,46 +440,51 @@ public void onConnection(Connection connection) {
+        if (closeConnectionReason.isIntended)
+            return;
+
+        if (!connection.getPeersNodeAddressOptional().isPresent())
```
Is there a reason why you prefer using directly `getPeersNodeAddressOptional().isPresent()` in comparison using `hasPeerNodeAddress()` ? It is used mixed in the original code as well, so we might use one way for the future.

julianknutsen replied:
I wish there was a better reason than this one, but I used this pattern because it reduces the number of lint errors from IDEA. If you don't have an `isPresent()` call prior to a `get()` it will highlight the word and raise an alert. You can suppress it, but that is just more code to add.

The other benefit is that mocking is bit less error-prone because you just need to mock out the Optional and not the wrapper class that has the `hasPeerNodeAddress()` function.

ripcurlx commented on p2p/src/main/java/bisq/network/p2p/storage/P2PDataStorage.java:
```diff
@@ -440,46 +440,51 @@ public void onConnection(Connection connection) {
+            // We only set the data back by half of the TTL and remove the data only if is has
+            // expired after that back dating.
+            // We might get connection drops which are not caused by the node going offline, so
+            // we give more tolerance with that approach, giving the node the change to
```
NIT: Small typo in the original code: `change` should be `chance`

julianknutsen replied:
Addressed in e8c8225

ripcurlx commented on p2p/src/main/java/bisq/network/p2p/storage/P2PDataStorage.java:
```diff
@@ -448,43 +448,24 @@ public void onDisconnect(CloseConnectionReason closeConnectionReason, Connection
+                    // We only set the data back by half of the TTL and remove the data only if is has
+                    // expired after that back dating.
+                    // We might get connection drops which are not caused by the node going offline, so
+                    // we give more tolerance with that approach, giving the node the change to
```
NIT: The `change` != `chance` typo in the original code is copied over to this place as well.

julianknutsen replied:
Addressed in e8c8225

ripcurlx reviewed (dismissed):
ACK

- Tested on local Regtest trading with each peer going offline after each step.
- Tested on local Regtest trader chat with each peer going offline after each message.

julianknutsen committed:
[PR COMMENTS] Fix comment typo

s/change/chance/

julianknutsen commented:
@ripcurlx This PR is in a strange GitHub state now. It has the "waiting for author" tag, but since it was ACKd and not "Request Changes" I don't have a button to push to move it back to your queue and can't remove the label myself. Not sure the right way to handle this type of workflow, but I wanted to give you a heads up.

In any event, I've addressed your comments and it is ready for another look.

ripcurlx commented:
> @ripcurlx This PR is in a strange GitHub state now. It has the "waiting for author" tag, but since it was ACKd and not "Request Changes" I don't have a button to push to move it back to your queue and can't remove the label myself. Not sure the right way to handle this type of workflow, but I wanted to give you a heads up.
> 
> In any event, I've addressed your comments and it is ready for another look.

It was probably not perfectly handled by myself, as I probably should have asked for change requested as status. In that case I was ok about everything, just had optional remarks.

Labels can only be changed by maintainers or triage permission owners. I'll have a look how to do it with GitHub actions automatically based on the PR state.

ripcurlx reviewed (approved):
ACK
